### ISSUE TITLE
Editing post type when site has more than 387 posts causes 503 with WP Advanced Page Cache enabled

### ISSUE THREAD

**Issue by ttrowell** [pantheon-systems/pantheon-advanced-page-cache#88]
Accessing the URL gives out 503 when WP Advanced Page Cache plugin is enabled. 
Deactivating the plugin enable the admin to load the page.
Error when WP-APC is enabled:
Error 503 Backend fetch failed
Backend fetch failed
Guru Meditation:
XID: 664770472
Varnish cache server

Additional info. The Dev site is using:
PHP 7.0.13-20161121
WP Core 4.7

```
WAS able to reproduce using the provided link:
https://xxxxxx-health.pantheonsite.io/wp-admin/edit.php?post_type=crb_news
WAS able to reproduce with a similar link with additional selected fields:
https://xxxxxxx-health.pantheonsite.io/wp-admin/edit.php?post_status=all&post_type=crb_news&x=y
Was NOT able to reproduce with same URL when additional content filtered:
https://xxxxxxxx-health.pantheonsite.io/wp-admin/edit.php?post_status=all&post_type=crb_news&m=201503&seo_filter&filter_action=Filter&paged=1
Was NOT able to reproduce when any other post_type was selected.
Was NOT able to reproduce with a clean WP install.
I tried accessing the site directly with nginx, bypassing varnish; however, the site did not render well, and I was not able to log in, so I wasn't able to run any tests like this.
On requests that result in an error 503:
Nginx access log DOES show that results are being requested from WordPress, and the error logs for Nginx, php-error and php-fpm-error DO NOT contain any error messages.
Investigating Varnish:
=================
https://getpantheon.atlassian.net/wiki/display/VULCAN/Useful+Varnish+tips
Find out which styx cluster the site is using:
$ dig xxxxxxxx-health.pantheonsite.io

; <<>> DiG 9.10.3-P3-RedHat-9.10.3-8.P3.fc22 <<>> xxxxxxxx-health.pantheonsite.io
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 41682
;; flags: qr rd ra; QUERY: 1, ANSWER: 4, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;xxxxxxxx-health.pantheonsite.io. IN A

;; ANSWER SECTION:
xxxxxxxxx-health.pantheonsite.io. 600 IN CNAME styx-01.pantheon.io.
styx-01.pantheon.io.	392	IN	A	104.239.201.60
styx-01.pantheon.io.	392	IN	A	104.239.201.70
styx-01.pantheon.io.	392	IN	A	23.253.58.44

Log in to styx:
$ ssh styx-01.pantheon.io
$ sudo varnishadm
varnish> panic.show
This showed that the last panic happened on 1 Jan 2017 at 7s past midnight; another server had a panic on the same day a little more than an hour after midnight. It seems that the 503 errors are not showing up as panic messages in styx, to the best I can tell. This conclusion might be in error, as the "Guru meditation" error message pretty clearly indicates a segfault, and I am not very familiar with varnish.
More


```

**Comment by ttrowell** [comment]
We have another customer complaint about this and I was able to reproduce in a multidev.
I've disabled all plugins except for WP Advanced Page Cache.

I also switched to one of the default themes.
This is happening on /wp-admin/edit.php?post_type=page, so not a custom post type.
I updated the plugin to the latest version (0.1.3) on the multidev.

**Comment by ttrowell** [comment]
I tested using good ol' print_r and it seems to be failing at the line below in inc/class-emitter.php. There seem to be lots of entries in the $keys variable.
My testing url was https://xxxxxxx.pantheonsite.io/wp-admin/edit.php?post_type=page
$keys = apply_filters( 'pantheon_wp_main_query_surrogate_keys', $keys );
Splitting the array just before the filter ($keys = array_slice($keys, 0, 350);), from 590 keys to 350 - no error. It fails with 400 keys though.

**Comment by ttrowell** [comment]
I have done some extensive testing and I found a pattern:

The 503 "backend fetch failed" error when viewing the page list only occurs when there are more than 387 pages in the chosen list.

For example:
When viewing "all," the combined number of drafts and published pages cannot exceed 387. When viewing "published," the number of published pages cannot exceed 387.
However, if there are 387 published pages and one draft page, `edit.php?post_status=publish&post_type=page` will display fine, but `edit.php?post_type=page&all_posts=1` will return the 503 error.

**Comment by ttrowell** [comment]
I wasn't able to replicate on freshly installed site. I tried a few custom post types, the one called "news" in particular. I added about 800 posts to it, taxonomy terms, comments, etc. I even tried setting the admin page to list all 800 of them with no luck. There could be some combination of plugins or scenario with these sites that are making them crash but I haven't been able to find it with a bare install.

**Comment by danielbachhuber** [comment]
@ttrowell @newtoid @joshkoenig My hunch is that the 503 error is coming from the call to `pantheon_clear_edge_keys()`, e.g. the underlying subsystem (Varnish) fails when too many surrogate keys (greater than 350-400) are passed to it.

If this is the case, PAPC could be failing intermittently on some WordPress installs when some cache clear action generates too many surrogate keys.

The easy fix would be to split the number of purged surrogate keys by some magic threshold. The more correct fix would be to perform similar behavior within `pantheon_clear_edge_keys()`

**Comment by danielbachhuber** [comment]
> My hunch is that the 503 error is coming from the call to `pantheon_clear_edge_keys()`, e.g. the underlying subsystem (Varnish) fails when too many surrogate keys (greater than 350-400) are passed to it.

This original hunch is invalid because I incorrectly assumed the 503 was experienced during a PURGE event. The reports indicate the 503 is experienced during a GET request that loads lots of posts (and generates corresponding surrogate keys)

I reproduced the issue by:

1. Creating a new test environment: http://dev-papc-test-may-2017.pantheonsite.io/
2. Generating 3500 posts with: `terminus wp papc-test-may-2017.dev -- post generate --count=3500`
3. Installing and activating PAPC v0.1.4
4. Adding the following to `functions.php`: `add_action( 'pre_get_posts', function( $query ) { $query->set( 'posts_per_page', -1 ); });`
4. Visiting an admin view that loaded all generated posts: http://dev-papc-test-may-2017.pantheonsite.io/wp-admin/edit.php

![image](https://cloud.githubusercontent.com/assets/36432/26423628/3a1f860a-4083-11e7-993c-65b6d4f2d2cd.png)

Essentially, WordPress + PAPC is generating too many surrogate keys, and it's exceeding the size Nginx can handle (~32kb)

The most obvious fix is that we shouldn't be emitting surrogate keys in the admin, so I'm going to ship a version that disables that, and we can see if it fixes the original reports.

**Comment by achbed** [comment]
I'm seeing this also, but in our case it's not just in the admin. We have a map that displays markers for 800+ hotels. When we output this, we break due to this issue.

**Comment by danielbachhuber** [comment]
Thanks for the report, @achbed.

> We have a map that displays markers for 800+ hotels. When we output this, we break due to this issue.

This makes sense. Given the existing thread, what would your preferred behavior be?

The obvious next step to me would be to cap the total emitted surrogate key size to 32kb. However, this opens us up to causing inexplicable cache purge failures (e.g. I updated a post but the page wasn't purged because there were too many posts on the page).

Another approach would be to change the post surrogate key from `post-<id>` to `p-<id>`. This would save us some bytes with every post but would be a breaking change. We could do the same change for terms too.

Open to other suggestions too. To fix your immediate issue, you can filter `pantheon_wp_main_query_surrogate_keys` to disable some or all surrogate keys from emitting on that view. If your URI is `/map/`, then the code snippet might look like this:

```
add_filter( 'pantheon_wp_main_query_surrogate_keys', function( $keys ) {
	if ( 0 === stripos( $GLOBALS['wp']->request, 'map' ) ) {
		return array();
	}
	return $keys;
});
```

**Comment by achbed** [comment]
I wasn't sure what we could do here, given the goal (selective cache purging) and the limitation (huge number of posts that overwhelms the header limits). The issue is that there doesn't seem to be a general case where this could be fixed.

This seems to be a fairly edge case. We might look at length, and if the total length is over a set limit mark it with a "posthuge" tag instead of all the individuals. We could add that tag to all clear requests. Hopefully that would limit impact.

**Comment by danielbachhuber** [comment]
> This seems to be a fairly edge case.

I agree.

> We might look at length, and if the total length is over a set limit mark it with a "posthuge" tag instead of all the individuals. We could add that tag to all clear requests.

This is an interesting idea. I'll discuss with @joshkoenig and follow up when there's a decision.

**Comment by achbed** [comment]
Proposed patch is at https://github.com/pantheon-systems/pantheon-advanced-page-cache/pull/94

**Comment by achbed** [comment]
Just curious if you and @joshkoenig had a chance to discuss this yet.

**Comment by danielbachhuber** [comment]
@achbed Yes, and we're generally onboard with the approach. I'm out with a fever today. I'll review again when I have a clear head.

**Comment by danielbachhuber** [comment]
This is permanently addressed with #94
